test/EtomicSwap.js

```diff
@@ -47,593 +47,663 @@ const invalidSecretHex = '0x' + invalidSecret.toString('hex');
 
-        // not allow to refund invalid amount
-        await this.swap.senderRefund(id, web3.utils.toWei('2'), secretHash, this.token.address, accounts[1]).should.be.rejectedWith(EVMThrow);
+        // Not allow to refund invalid amount
+        await etomicSwap.connect(accounts[0]).senderRefund(id, ethers.parseEther('2'), secretHash, token.target, accounts[1].address).should.be.rejected;
 
-        // success refund
-        const balanceBefore = web3.utils.toBN(await this.token.balanceOf(accounts[0]));
+        // Success refund
+        const balanceBefore = await token.balanceOf(accounts[0].address);
 
-        await this.swap.senderRefund(id, web3.utils.toWei('1'), secretHash, this.token.address, accounts[1]).should.be.fulfilled;
+        await etomicSwap.connect(accounts[0]).senderRefund(id, ethers.parseEther('1'), secretHash, token.target, accounts[1].address).should.be.fulfilled;
 
-        const balanceAfter = web3.utils.toBN(await this.token.balanceOf(accounts[0]));
+        const balanceAfter = await token.balanceOf(accounts[0].address);
 
-        // check sender balance
-        assert.equal(balanceAfter.sub(balanceBefore).toString(), web3.utils.toWei('1'));
+        // Check sender balance
+        expect((balanceAfter - balanceBefore)).to.equal(ethers.parseEther('1'));
 
-        const payment = await this.swap.payments(id);
-        assert.equal(payment[2].valueOf(), SENDER_REFUNDED);
+        // Check the state of the payment
+        const payment = await etomicSwap.payments(id);
+        expect(payment.state).to.equal(BigInt(SENDER_REFUNDED));
 
-        // not allow to refund again
-        await this.swap.senderRefund(id, web3.utils.toWei('1'), secretHash, this.token.address, accounts[1]).should.be.rejectedWith(EVMThrow);
+        // Do not allow to refund again
+        await etomicSwap.connect(accounts[0]).senderRefund(id, ethers.parseEther('1'), secretHash, token.target, accounts[1].address).should.be.rejected;
     });
 
-    it('should allow sender to refund ERC721 payment after locktime', async function () {
+    it('should allow sender to refund ERC721 payment after locktime', async function() {
         const lockTime = await currentEvmTime() + 1000;
         const tokenId = 1;
 
-        const data = web3.eth.abi.encodeParameters(
+        const abiCoder = new AbiCoder();
+        const data = abiCoder.encode(
             ['bytes32', 'address', 'address', 'bytes20', 'uint64'],
-            [id, accounts[1], this.erc721token.address, secretHash, lockTime]
+            [id, accounts[1].address, erc721token.target, secretHash, lockTime]
         );
         // Call safeTransferFrom directly to transfer the token to the EtomicSwap contract
-        await this.erc721token.safeTransferFrom(accounts[0], this.swap.address, tokenId, data).should.be.fulfilled;
+        await erc721token['safeTransferFrom(address,address,uint256,bytes)'](accounts[0].address, etomicSwap.target, tokenId, data).should.be.fulfilled;
```

Comment:
Could you please clarify why you didn't use `connect` here?

Reply:
Oh I left `safeTransferFrom` without connection, as this function checks itself that `_from` address should be owner. I think accounts[0] is used by default, thats why transfer doesnt fail.

However, according `ethers.js` development practices using `connect` is essential when you want to perform a state-changing operation. So I added connect to NFT `safeTransferFrom` functions too.

Reply:
Ok, thanks. Why does it also require access to function through square brackets instead of `.safeTransferFrom`? :slightly_smiling_face:

Reply:
> Ok, thanks. Why does it also require access to function through square brackets instead of `.safeTransferFrom`? 🙂

The thing is that `ERC721` has two variants of `safeTransferFrom` methos, one requires `_data` param, another doesnt have it. You can find them in specification section https://eips.ethereum.org/EIPS/eip-721
 
```solidity
    /// @notice Transfers the ownership of an NFT from one address to another address
    /// @dev Throws unless `msg.sender` is the current owner, an authorized
    ///  operator, or the approved address for this NFT. Throws if `_from` is
    ///  not the current owner. Throws if `_to` is the zero address. Throws if
    ///  `_tokenId` is not a valid NFT. When transfer is complete, this function
    ///  checks if `_to` is a smart contract (code size > 0). If so, it calls
    ///  `onERC721Received` on `_to` and throws if the return value is not
    ///  `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`.
    /// @param _from The current owner of the NFT
    /// @param _to The new owner
    /// @param _tokenId The NFT to transfer
    /// @param data Additional data with no specified format, sent in call to `_to`
    function safeTransferFrom(address _from, address _to, uint256 _tokenId, bytes data) external payable;

    /// @notice Transfers the ownership of an NFT from one address to another address
    /// @dev This works identically to the other function with an extra data parameter,
    ///  except this function just sets data to "".
    /// @param _from The current owner of the NFT
    /// @param _to The new owner
    /// @param _tokenId The NFT to transfer
    function safeTransferFrom(address _from, address _to, uint256 _tokenId) external payable;
```

Without `safeTransferFrom` clarification in hardhat env there will be the error
```
TypeError: ambiguous function description (i.e. matches "safeTransferFrom(address,address,uint256)", "safeTransferFrom(address,address,uint256,bytes)") (argument="key", value="safeTransferFrom", code=INVALID_ARGUMENT, version=6.10.0)
```

Reply:
I see, thanks for the info :slightly_smiling_face: